README.md
# AI workshop
All instructions for the workshop can be found in the ai_biodiv_tutorial file. You can open it either as a jupyter notebook or as a html file in your browser (download file first). To run the tutorial you will need python and preferably also conda installed. Follow instructions below to install.

## Installing and setting up Python

Here are a few steps of preparations to make sure you will all be able to run Python on your personal computers and install any required packages. If you already have a working Python editor setup on your computer, just ignore these instructions.

1. Please install **minconda** on your computer (unless you already have some conda version installed): https://docs.conda.io/en/latest/miniconda.html
2. Once installed you will be able to run the conda installation manager through your command line to install software with conda (see `conda install` commands below). Mac and Linux users can just use the native **Terminal** app, but Windows users need to make sure to use the **conda command prompt** that should install together with miniconda. If you as a Windows user can't find the conda command prompt or it didn't install for some reason, you can download anaconda instead (https://www.anaconda.com/products/individual)
3. Now it's time to install some Python editor. You can choose to either work inside a **jupyter notebook** or within a **Python editor** (similar to R-Studio). Don't worry if you don't know at this point what either of these are, but install one of them. I recommend a Python editor for a better learning effect.

*Jupyter notebook:*
- conda install jupyter
- alternatively you can download Visual Studio Code (https://code.visualstudio.com) and install the jupyter plugin (https://code.visualstudio.com/docs/datascience/jupyter-notebooks)

*Python studio editor:*
- conda install spyder
- alternatively you can download PyCharm (https://www.jetbrains.com/pycharm/)


data/history.txt
loss:0.11085620522499084,0.08415737748146057,0.07679077237844467,0.07368314266204834,0.07204087823629379,0.06856495887041092,0.06713633239269257,0.06628245115280151,0.06397052109241486,0.06334356218576431,0.0611419714987278,0.060214508324861526,0.059111498296260834,0.0599704273045063,0.05898692458868027,0.05920254439115524,0.05725039169192314,0.056613463908433914,0.05607360601425171,0.05657397210597992,0.055195849388837814,0.05537506192922592,0.054900795221328735,0.054989639669656754,0.05382692441344261,0.054421428591012955,0.05418778210878372,0.052883345633745193,0.053012702614068985,0.05267663672566414,0.052570272237062454,0.052962422370910645,0.0518568754196167,0.052001338452100754,0.051945656538009644,0.05098619684576988,0.05071745440363884,0.05047610029578209,0.05118624120950699,0.05128953233361244,0.05060882866382599,0.04989253357052803,0.050510525703430176,0.049588531255722046,0.049470096826553345,0.05009782686829567,0.04918786883354187,0.04945911839604378,0.04934366047382355,0.049150921404361725,0.04940386489033699,0.049050260335206985,0.048712048679590225,0.049734167754650116,0.048354748636484146,0.0477047860622406,0.04812943562865257,0.04764728993177414,0.04796672612428665,0.048103831708431244,0.04828234761953354,0.04802204295992851,0.04827897250652313,0.0480765774846077,0.0480065792798996,0.047924768179655075,0.046946894377470016,0.04823099076747894,0.049726009368896484,0.04650444537401199,0.047245852649211884,0.04679522663354874,0.046987954527139664,0.04651286080479622,0.04647395387291908,0.046405911445617676,0.046661533415317535,0.04621555283665657,0.04627252370119095,0.0460045263171196,0.0478552021086216,0.04624079167842865,0.04632895439863205,0.046006519347429276,0.046902697533369064,0.046027351170778275,0.045698124915361404,0.04633698984980583,0.04655926674604416,0.045612022280693054,0.04627228155732155,0.045982908457517624,0.04597923159599304,0.04666408151388168,0.04604330286383629,0.04615895077586174,0.04586935043334961,0.04525166004896164,0.045363329350948334,0.045395378023386,0.04561666399240494,0.04494289681315422,0.04514739289879799,0.04584372043609619,0.04532737657427788,0.04575498402118683,0.04553134739398956,0.044743284583091736,0.044724538922309875,0.04523782804608345,0.044366877526044846,0.045217081904411316,0.04480992257595062,0.04481177404522896,0.04477465897798538,0.04461219161748886,0.04459041357040405,0.044171012938022614,0.04427037760615349,0.04519526660442352,0.04506959766149521,0.044509854167699814,0.04490642249584198,0.04462990537285805,0.046135593205690384,0.04503045976161957,0.04398585110902786,0.04355308786034584,0.043750349432229996,0.04421692341566086,0.043068092316389084,0.043174006044864655,0.043990038335323334,0.04387950524687767,0.04372023418545723,0.04483911022543907,0.04364609718322754,0.043703410774469376,0.04326530173420906,0.043252017349004745,0.04279699549078941,0.04351174831390381,0.04311079904437065,0.042941197752952576,0.04398506507277489,0.04310047999024391,0.04272330179810524,0.04363070800900459,0.04371831193566322,0.04281315952539444,0.042261380702257156,0.04264475032687187,0.04241194948554039,0.042419783771038055,0.04298115149140358,0.04265109449625015,0.04322914034128189,0.04346780851483345,0.04256696626543999,0.04296783730387688,0.04245561361312866,0.04257401451468468,0.042516790330410004,0.042780425399541855,0.0430799201130867,0.04231715574860573,0.042393531650304794,0.04415559396147728,0.04220345988869667,0.04218544811010361,0.04167097061872482,0.041874609887599945,0.042023081332445145,0.04249952733516693,0.04272184148430824,0.04204820469021797,0.041718486696481705,0.04232519492506981,0.04285898059606552,0.04244948551058769,0.041346367448568344,0.042367737740278244,0.041667867451906204,0.04159809648990631,0.042056843638420105,0.04206215217709541,0.04115966707468033,0.04157879576086998,0.041879646480083466,0.041422076523303986,0.042662687599658966,0.042465075850486755,0.04115420952439308,0.041256681084632874,0.04126590117812157,0.041688643395900726,0.041908275336027145,0.041801583021879196,0.04162409529089928,0.04210551455616951,0.04109425097703934,0.041282348334789276,0.04140676185488701,0.04196438938379288,0.04183158650994301,0.04186609759926796,0.041379738599061966,0.04112327843904495,0.04102124273777008,0.04135783016681671,0.04104727506637573,0.04087624326348305,0.04084848612546921,0.04041437804698944,0.04060603678226471,0.04132017120718956,0.04109228774905205,0.04076467826962471,0.04071081802248955,0.0417884886264801,0.04114373028278351,0.04082464054226875,0.040730252861976624,0.041009221225976944,0.04097446799278259,0.04099300503730774,0.040611542761325836,0.041426874697208405,0.04126814752817154,0.04076920822262764,0.040769610553979874,0.04097379744052887,0.04122190922498703,0.04065096378326416,0.040640465915203094,0.04064131900668144,0.040632959455251694,0.04027855396270752,0.04114300385117531,0.04023332521319389,0.040694404393434525,0.040298666805028915,0.04016320779919624,0.04038289189338684,0.04064958542585373,0.04000946506857872,0.040150661021471024,0.04065028205513954,0.03990570083260536,0.040128689259290695,0.04030716046690941,0.04135628789663315,0.04029213264584541,0.04000164568424225,0.03979923576116562,0.03989718481898308,0.039783019572496414,0.03981192409992218,0.04036888852715492,0.040818654000759125,0.04001360386610031,0.03951745480298996,0.040719423443078995,0.03987149894237518,0.03929709270596504,0.040169667452573776,0.0394611693918705,0.040273331105709076,0.03951694071292877,0.04100032523274422,0.04035899043083191,0.039848845452070236,0.039885737001895905,0.03961292281746864,0.039831534028053284,0.03931931406259537,0.03960416465997696,0.03909644857048988,0.040158357471227646,0.03943478316068649,0.03912980854511261,0.03974087908864021,0.04002027213573456,0.03985074907541275,0.03914700448513031,0.039324138313531876,0.03996129706501961,0.03931436687707901,0.03965110331773758,0.03876199573278427,0.0398220531642437,0.03933950141072273,0.03913082182407379,0.039394963532686234,0.039123114198446274,0.038984693586

src/test_cgr_dna.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Dec  3 16:22:28 2020

@author: Tobias Andermann (tobiasandermann88@gmail.com)
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import collections
from matplotlib import cm

def count_kmers(sequence, k):
    d = collections.defaultdict(int)
    for i in range(len(sequence)-(k-1)):
        d[sequence[i:i+k]] +=1
    for key in d.keys():
        if "N" in key:
            del d[key]
    return d

def probabilities(sequence, kmer_count, k):
    probabilities = collections.defaultdict(float)
    N = len(sequence)
    for key, value in kmer_count.items():
        probabilities[key] = float(value) / (N - k + 1)
    return probabilities

def chaos_game_representation(probabilities, k):
    array_size = int(np.sqrt(4**k))
    chaos = np.zeros([array_size,array_size])
    maxx = array_size
    maxy = array_size
    posx = 1
    posy = 1
    for key, value in probabilities.items():
        for char in key:
            if char == "T":
                posx += maxx / 2
            elif char == "C":
                posy += maxy / 2
            elif char == "G":
                posx += maxx / 2
                posy += maxy / 2
            maxx = maxx / 2
            maxy /= 2
        chaos[int(posy)-1][int(posx)-1] = value
        maxx = array_size
        maxy = array_size
        posx = 1
        posy = 1
    return chaos

dna_seq = 'ACTAGTATTCGCGCTCCTTGGAGCGGTGCATGGTCTTGACAAAATATGTCTTGGACATCATGCAGTCTCCAACGGCACCATCGTAAAGACTCTCACAAACGAAAAGGAAGAGGTGACCAATGCTACTGAAACAGTGGAGAGCAAAAGCCTGGACAAACTTTGCATGAAAAGTCGAAATTATAAGGACTTAGGTAATTGCCACCCAATAGGGATGATAATAGGGACTCCTGCTTGTGATCTACATCTTACCGGAACATGGGACACTTTAATAGAGAGAGACAATTCTATTGCCTATTGTTACCCGGGTGCCACTGTGAATGAAGAAGCATTAAGGCAGAAAATTATGGAAAGTGGAGGGATTGACAAGATAAGCACCGGGTTCACATATGGATCATCCATCAATTCAGCTGGAACCACTAAAGCATGTATGAGAAATGGAGGAAATAGTTTCTATGCGGAACTAAAGTGGCTAGTGTCGAAAAGCAAAGGACAAAACTTCCCACAAACAACAAACACATACAGAAATACAGATTCAGCAGAACATCTTATAATCTGGGGAATTCATCACCCTTCAAGCACACAAGAAAAGAATGACCTATATGGAACACAATCACTTTCCATTTCAGTGGGGAGTTCTACTTATCAAAACAACTTTGTGCCTGTGGTGGGAGCAAGACCACAAGTGAATGGCCAAAGCGGGCGGATTGACTTCCATTGGACGATGGTACAACCAGGTGATAACATCACTTTCTCGCATAATGGTGGATTAATAGCACCCAGCAGAGTGAGTAAGCTAAAGGGAAGAGGCCTTGGCATTCAATCAGGAGCTTCAGTAGACAATGACTGTGAGTCAAAATGTTTCTGGAAAGGTGGATCCATCAACACCAAACTCCCTTTTCAGAATCTTTCTCCAAGAACCGTGGGTCAATGCCCCAAGTATGTGAACAAAAAGAGCCTGTTGCTTGCTACTGGAATGAGGAATGTGCCAGAGGTTGTCCAAGGAAGAGGCCTGTTTGGAGCAATAGCTGGATTCATAGAAAATGGATGGGAAGGGATGGTAGATGGTTGGTATGGTTTCCGGCACCAAAATGCCCAAGGCACTGGTCAGGCCGCGGATTACAAAAGTACTCAGGCAGCTATAGATCAAATAACCGGGAAATTGAACAGGCTGATAGAGAAGACAAACACAGAATTCGAATCCATAGAGTCTGAGTTCAGTGAAATAGAACATCAAATCGGTAATGTAATAAACTGGACTAAAGATTCTATAACAGACATCTGGACGTATCAAGCTGAATTGCTGGTGGCAATGGAAAATCAGCATACAATTGACATGGCTGATTCAGAAATGCTGAATCTATATGAGAGAGTGAGGAAACAACTAAGGCAAAATGCTGAAGAAGATGGGAAAGGGTGCTTTGAGATATATCACAAATGCGATGACAACTGCATGGAAAGCATCAGAAACAACACTTATGACCATACACAATACAGAGAAGAAGCACTCTTGAACAGACTCAACATTAATCCGGTGAAACTCTCTTCTGGGTACAAAGATGTTATACTATGGTTCAGCTTCGGGGCGTCATGCTTTGTACTTTTGGCTGTCATCATGGGGCTTGTTTTCTTCTGCTTGAAAAATGGAAACATGCGATGCACAATCTGTATTTAGTTAAAAACACC'
kmer_size = 4
kmer_count = count_kmers(dna_seq,kmer_size)
kmer_probs = probabilities(dna_seq, kmer_count, kmer_size)
cgr_array = chaos_game_representation(kmer_probs, kmer_size)


plt.imshow(cgr_array, interpolation='nearest', cmap=cm.gray_r)
plt.show()



src/biodiv_nn_model.py
import numpy as np
import pandas as pd
import tensorflow as tf
import matplotlib.pyplot as plt

# This is a function that let's you rescale your data array. It rescales all values in each column to a range between 0 and 1.
def min_max_scaler(x, min_array=None, max_array=None, inverse_scale=False):
    if inverse_scale:
        x_new = x*(max_array-min_array)+min_array
        return(x_new)
    else:
        min_array = np.min(x,axis=0)
        max_array = np.max(x,axis=0)
        x_new = (x - min_array)/(max_array - min_array)
        return(x_new,min_array,max_array)


def select_train_val_test(x,val_fraction=0.2,test_fraction=0.2,shuffle=True,seed=None):
    all_indices = np.arange(len(x))
    if shuffle:
        if not seed:
            seed = np.random.randint(0,999999999)
        # shuffle all input data and labels
        np.random.seed(seed)
        print('Shuffling data, using seed', seed)
        shuffled_indices = np.random.choice(all_indices, len(all_indices), replace=False)
    else:
        shuffled_indices = all_indices
    # select train, validation, and test data
    n_test_instances = np.round(len(shuffled_indices) * test_fraction).astype(int)
    n_validation_instances = np.round(len(shuffled_indices) * val_fraction).astype(int)
    test_ids = shuffled_indices[:n_test_instances]
    validation_ids = shuffled_indices[n_test_instances:n_test_instances + n_validation_instances]
    train_ids = shuffled_indices[n_test_instances + n_validation_instances:]
    return train_ids, validation_ids, test_ids

# class MCDropout(tf.keras.layers.Dropout):
#     def call(self, inputs):
#         return super().call(inputs, training=True)

def plot_training_history(history, show_best_epoch=True):
    fig = plt.figure(figsize=(8, 5))
    plt.plot(history.history['mae'], label='Training set')
    plt.plot(history.history['val_mae'], label='Validation set')
    plt.xlabel('Epochs')
    plt.ylabel('MAE')
    if show_best_epoch:
        best_epoch = np.where(history.history['val_mae'] == np.min(history.history['val_mae']))[0][0]
        plt.axvline(best_epoch, c='grey', linestyle='--')
        plt.axhline(history.history['val_mae'][best_epoch], c='grey', linestyle='--')
        plt.gca().axvspan(best_epoch, len(history.history['mae']), color='grey', alpha=0.3, zorder=3)
    plt.grid()
    plt.legend(loc='upper center')
    plt.show()

def plot_true_vs_pred(true_labels, predicted_labels):
    fig = plt.figure(figsize=(6, 6))
    plt.plot(true_labels, predicted_labels, 'o', markersize=3, alpha=1)
    plt.grid()
    plt.xlabel('True diversity')
    plt.ylabel('Predicted diversity')
    plt.show()


# load data
tbl = pd.read_csv("/Users/tobiasandermann/Documents/teaching/ai_course_geosciences/biodiv_exercise/data/div_data_all_features.txt", delimiter="\t")
features = tbl.values[:,1:]
feature_names = tbl.columns[1:]
labels = tbl.values[:,0]
plt.hist(labels)
plt.show()
# Rescale labels and features
rescaled_labels = labels/800
rescaled_features, scale_min, scale_max = min_max_scaler(features)
#min_max_scaler(rescaled_features, scale_min, scale_max,inverse_scale=True)

# Separate instances into train and test set
train_set_ids, validation_set_ids, test_set_ids =  select_train_val_test(rescaled_features)
train_features = rescaled_features[train_set_ids]
train_labels = rescaled_labels[train_set_ids]
validation_features = rescaled_features[validation_set_ids]
validation_labels = rescaled_labels[validation_set_ids]
test_features = rescaled_features[test_set_ids]
test_labels = rescaled_labels[test_set_ids]


architecture = []
# Input layer
architecture.append(tf.keras.layers.Flatten(input_shape=[train_features.shape[1]]))
# First hidden layer
architecture.append(tf.keras.layers.Dense(32, activation='relu'))
architecture.append(tf.keras.layers.Dropout(0.2))
#architecture.append(MCDropout(0.2))
# 2nd hidden layer
architecture.append(tf.keras.layers.Dense(8, activation='relu'))
architecture.append(tf.keras.layers.Dropout(0.1))
#architecture.append(MCDropout(0.1))
# Output layer
architecture.append(tf.keras.layers.Dense(1, activation='softplus'))  # sigmoid or tanh or softplus
# Compile the model
model = tf.keras.Sequential(architecture)
model.compile(loss='mae', optimizer='adam', metrics=['mae','mape','mse','msle'])
# Get overview of model architecture
model.summary()



architecture = []

# Input layer
architecture.append(tf.keras.layers.Flatten(input_shape=[4]))
# 1st hidden layer
architecture.append(tf.keras.layers.Dense(5, activation='relu',use_bias=False))
# Output layer
architecture.append(tf.keras.layers.Dense(5, activation='softmax',use_bias=False))

# Compile the model
model = tf.keras.Sequential(architecture)
model.compile(loss='categorical_crossentropy',
 optimizer='adam', metrics=['accuracy'])

# Get overview of model architecture
model.summary()



# Define early stop of training based on validation set
early_stop = tf.keras.callbacks.EarlyStopping(monitor='val_mae',patience=200,restore_best_weights=True)
# Run model training and store training history
history = model.fit(train_features,
                          train_labels,
                          epochs=1000,
                          validation_data=(validation_features, validation_labels),
                          verbose=1,
                          callbacks=[early_stop],
                          batch_size=40)

plot_training_history(history,show_best_epoch=False)

# You can save the model to a file and later load it
model_file = '/Users/tobiasandermann/Documents/teaching/ai_course_geosciences/biodiv_exercise/data/trained_model'
model.save(model_file)
# This is how you load the model
#model = tf.keras.models.load_model(model_file)

estimated_train_labels = model.predict(train_features)
estimated_validation_labels = model.predict(validation_features)
estimated_test_labels = model.predict(test_features)




plot_true_vs_pred(train_labels*800,estimated_train_labels*800)
plot_true_vs_pred(validation_labels*800,estimated_validation_labels*800)
plot_true_vs_pred(test_labels*800,estimated_test_labels*800)




mc_dropout_pred = np.stack([model(test_features,training=True) for i in np.arange(100)])
mc_dropout_mean = mc_dropout_pred.mean(axis=0)
mc_dropout_std = mc_dropout_pred.std(axis=0)
mc_dropout_min = mc_dropout_pred.min(axis=0)
mc_dropout_max = mc_dropout_pred.max(axis=0)

plot_true_vs_pred(test_labels*800,mc_dropout_mean*800)


instance_id = 4
stepsize = 5
estimates = mc_dropout_pred[:,instance_id,:].flatten()*800
estimates = np.random.normal(2.1,0.2,100)
plt.hist(estimates,20)#np.arange(min(estimates),max(estimates+stepsize),stepsize))
plt.axvline(np.mean(estimates), color='r', linestyle='dashed', linewidth=1)
plt.xlim(0,5.5)
plt.grid()
plt.show()

np.log(10)

fig = plt.figure(figsize=[5,5])
#plt.errorbar(test_labels*800,mc_dropout_mean*800, yerr=[mc_dropout_min.flatten()*800,mc_dropout_max.flatten()*800], fmt='o',ecolor='black',elinewidth=0.2)
plt.errorbar(test_labels*800,mc_dropout_mean*800, yerr=mc_dropout_std.flatten()*800, fmt='.',alpha=1,ecolor='black',elinewidth=0.5)
plt.xlim(0,800)
plt.ylim(0,800)
plt.plot([0,800],[0,800],'r-')
plt.xlabel('True diversity')
plt.ylabel('Predicted diversity')
plt.tight_layout()
fig.savefig('/Users/tobiasandermann/Documents/teaching/ai_course_geosciences/biodiv_exercise/figs/mc_dropout_pred.pdf',bbox_inches='tight', dpi = 500)


import os
indir = '/Users/tobiasandermann/Documents/teaching/ai_course_geosciences/biodiv_exercise/data/iucn_data/'
labels_path = os.path.join(indir,'iucnn_train_labels.txt')
a = pd.read_csv(labels_path,sep='\t')
b = a['labels']
b.to_csv(labels_path,sep='\t',index=False)


src/turn_sequences_to_img.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Dec  3 17:53:10 2020

@author: Tobias Andermann (tobiasandermann88@gmail.com)
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from Bio import SeqIO
import collections
from matplotlib import cm
import os

def count_kmers(sequence, k):
    d = collections.defaultdict(int)
    for i in range(len(sequence)-(k-1)):
        d[sequence[i:i+k]] +=1
    for key in d.keys():
        if "N" in key:
            del d[key]
    return d

def probabilities(sequence, kmer_count, k):
    probabilities = collections.defaultdict(float)
    N = len(sequence)
    for key, value in kmer_count.items():
        probabilities[key] = float(value) / (N - k + 1)
    return probabilities

def chaos_game_representation(probabilities, k):
    array_size = int(np.sqrt(4**k))
    chaos = np.zeros([array_size,array_size])
    maxx = array_size
    maxy = array_size
    posx = 1
    posy = 1
    for key, value in probabilities.items():
        for char in key:
            if char == "T":
                posx += maxx / 2
            elif char == "C":
                posy += maxy / 2
            elif char == "G":
                posx += maxx / 2
                posy += maxy / 2
            maxx = maxx / 2
            maxy /= 2
        chaos[int(posy)-1][int(posx)-1] = value
        maxx = array_size
        maxy = array_size
        posx = 1
        posy = 1
    return chaos


# read all the sequences
input_file = '/Users/xhofmt/GitHub/dna_sequence_nn/data/sequence.fasta'
genus_sequence_dict = {}
for record in SeqIO.parse(input_file,'fasta'):
    genus = record.description.replace(record.id+' ','').split(' ')[0]
    if not genus.isupper():
        pass
    genus_sequence_dict.setdefault(genus,[])
    genus_sequence_dict[genus].append(str(record.seq))


# select the best taxa (those with most sequences)
produce_n_images = 30
n_seqs = np.array([len(genus_sequence_dict[i]) for i in genus_sequence_dict])
taxa = np.array([i for i in genus_sequence_dict])
#n_seqs[n_seqs>=produce_n_images]
selected_taxa = taxa[n_seqs>=produce_n_images]
#index_order = np.argsort(n_seqs)[::-1]
#taxa_sorted = np.array(taxa)[index_order]
final_taxon_selection = [i for i in selected_taxa if i not in ['Homo', 'UNVERIFIED:']]


# create DNA images for selected taxa
out_folder = '/Users/xhofmt/GitHub/dna_sequence_nn/data/images'
train_out_folder = os.path.join(out_folder,'train')
validation_out_folder = os.path.join(out_folder,'validation')
kmer_size = 4
for taxon in final_taxon_selection:
    seqs = np.random.choice(genus_sequence_dict[taxon],produce_n_images,replace=False)
    for i,dna_seq in enumerate(seqs):
        if i < int(np.round(0.66*produce_n_images)):        
            taxon_out = os.path.join(train_out_folder,taxon)
        else:
            taxon_out = os.path.join(validation_out_folder,taxon)
        if not os.path.exists(taxon_out):
            os.makedirs(taxon_out)
        dna_seq = dna_seq.replace('N','')
        kmer_count = count_kmers(dna_seq,kmer_size)
        kmer_probs = probabilities(dna_seq, kmer_count, kmer_size)
        cgr_array = chaos_game_representation(kmer_probs, kmer_size)
        f = plt.figure(figsize=[5,5])
        plt.imshow(cgr_array, interpolation='nearest', cmap=cm.gray_r)      
        plt.gca().set_axis_off()
        plt.margins(0,0)
        plt.gca().xaxis.set_major_locator(plt.NullLocator())
        plt.gca().yaxis.set_major_locator(plt.NullLocator())
        plt.savefig(os.path.join(taxon_out,'%04d.png'%i), bbox_inches = 'tight',pad_inches = 0)
        f.clear()
        plt.close(f)



src/apply_nn_image_classifier.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Dec  3 17:25:51 2020

@author: Tobias Andermann (tobiasandermann88@gmail.com)
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Sequential
from keras.layers import Conv2D
from keras.layers import MaxPooling2D
from keras.layers import Flatten
from keras.layers import Dense
import os, glob
os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE"


def make_convnet(instance,n_classes):
    input_shape = instance.shape
    model = Sequential()
    model.add(Conv2D(32, kernel_size=(5, 5), strides=(1, 1),activation='relu',input_shape=input_shape))
    model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
    model.add(Conv2D(64, (5, 5), activation='relu'))
    model.add(MaxPooling2D(pool_size=(2, 2)))
    model.add(Flatten())
    model.add(Dense(1000, activation='relu'))
    model.add(Dense(500, activation='relu'))
    model.add(Dense(200, activation='relu'))
    model.add(Dense(n_classes, activation='sigmoid'))
    model.compile(optimizer='Adam',loss='binary_crossentropy',metrics=['accuracy'])
    return model



data_gen = ImageDataGenerator(rescale=1.0/255)

imgdir = '/Users/xhofmt/GitHub/dna_sequence_nn/data/images'
img_size = 16
batch_size = 5
classes = [os.path.basename(i) for i in glob.glob(os.path.join(imgdir,'train/*'))]
n_classes = len(classes)


train_generator = data_gen.flow_from_directory(
        imgdir + '/train',
        target_size=(img_size, img_size),
        batch_size=batch_size,
        class_mode='categorical',
        classes=classes,
        shuffle=True)


Xbatch, Ybatch = train_generator.next()

np.shape(Xbatch)

plt.imshow(Xbatch[0])

    
my_cnn = make_convnet(Xbatch[0],n_classes)


validation_generator = data_gen.flow_from_directory(
        imgdir + '/validation',
        target_size=(img_size, img_size),
        batch_size=batch_size,
        class_mode='categorical',
        classes=classes,
        shuffle=False)



history = my_cnn.fit_generator(train_generator, epochs=100,validation_data=validation_generator,steps_per_epoch=len(train_generator),validation_steps=len(validation_generator),verbose=2)
my_cnn.save_weights('/Users/xhofmt/GitHub/dna_sequence_nn/trained_models/my_cnn_weights')
#my_cnn.load_weights('/Users/xhofmt/GitHub/dna_sequence_nn/trained_models/my_cnn_weights')

plt.plot(history.history['loss']);
plt.plot(history.history['val_loss']);
plt.legend(['training loss', 'validation loss']);


plt.plot(history.history['accuracy']);
plt.plot(history.history['val_accuracy']);
plt.legend(['training accuracy','validation accuracy']);




prediction = np.argmax(my_cnn.predict(validation_generator),axis=1)
true_labels = validation_generator.classes

sum(prediction==true_labels)/len(true_labels)


